### PR TITLE
Improves static validation tests

### LINKED ISSUE #413: Unit tests for validation on !isCreate
Follow-up from #404.

On a non-creating PUT, we don't currently test that:
* we don't care about the contents of WorkerProfiles or Version (as long as they don't change).
* we do care about the contents of every other field.

i.e. the `if isCreate`s added in #404 don't have test coverage, and I think they should.

### PR DESCRIPTION
**Pull request by m1kola**
Now we run tests for static validation for updates too.

Fixes #413

### CODE REVIEW

File: pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go [hunk 1]
```diff
@@ -88,58 +88,65 @@ func validOpenShiftCluster() *OpenShiftCluster {
...
-				(&openShiftClusterConverter{}).ToInternal(validOpenShiftCluster(), current)
-			}
-
-			err := v.Static(oc, current)
-			if err == nil {
-				if tt.wantErr != "" {
-					t.Error(err)
-				}
+func runTests(t *testing.T, isCreate bool, tests []*validateTest) {
+	variant := "Create"
```
Comment (m1kola) [inline pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go:92]: To review this function, I recommend hiding whitespace changes. I basically just changed arguments a bit and wrapped body with `t.Run(variant, func(t *testing.T) {...}`

File: pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go [hunk 2]
```diff
@@ -192,6 +200,8 @@ func TestOpenShiftClusterStaticValidateProperties(t *testing.T) {
 			},
 			wantErr: "400: InvalidParameter: properties.provisioningState: The provided provisioning state 'invalid' is invalid.",
 		},
+	}
+	createTests := append(commonTests, []*validateTest{
```
Comment (m1kola) [inline pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go:204]: Here we check validation of worker profiles (and below similarly we test version) only on create. However we make sure that it's not possible to change the worker profiles, version and other fields on update in `TestOpenShiftClusterStaticValidateDelta`.

File: pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go [hunk 3]
```diff
@@ -88,58 +88,65 @@ func validOpenShiftCluster() *OpenShiftCluster {
...
-			err := v.Static(oc, current)
-			if err == nil {
-				if tt.wantErr != "" {
-					t.Error(err)
-				}
+func runTests(t *testing.T, isCreate bool, tests []*validateTest) {
+	variant := "Create"
+	if !isCreate {
+		variant = "Update"
+	}
```
Comment (nilsanderselde) [inline pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go:95]: Wouldn't it be a little more straightforward to do the following?

```
variant = "Update"
if isCreate {
	variant = "Create"
}
```

File: pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go [hunk 4]
```diff
@@ -88,58 +88,65 @@ func validOpenShiftCluster() *OpenShiftCluster {
...
-				current = &api.OpenShiftCluster{}
-				(&openShiftClusterConverter{}).ToInternal(validOpenShiftCluster(), current)
-			}
-
-			err := v.Static(oc, current)
-			if err == nil {
-				if tt.wantErr != "" {
-					t.Error(err)
-				}
+func runTests(t *testing.T, isCreate bool, tests []*validateTest) {
```
Comment (jim-minter) [inline pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go:91]: I feel bad that I previously added `delta bool`.

`isCreate bool` is very opaque at point of use and then using a mixture of isCreate and variant in this function isn't that great either.  Create an enum type with two consts create and update, and use those throughout?

File: pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go [hunk 5]
```diff
@@ -206,22 +216,17 @@ func TestOpenShiftClusterStaticValidateProperties(t *testing.T) {
 			},
 			wantErr: "400: InvalidParameter: properties.workerProfiles: There should be exactly one worker profile.",
 		},
-	}
+	}...)
 
-	runTests(t, tests, false)
+	runTests(t, true, createTests)
+	runTests(t, false, commonTests)
```
Comment (jim-minter) [inline pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go:222]: I wonder whether the following might be a little clearer
```
commonTests := []*validateTest{...}
createTests := []*validateTest{...}

runTests(t, enumUpdate, commonTests)
runTests(t, enumCreate, commonTests)
runTests(t, enumCreate, createTests)
```

I mean, I actually had to check the Go spec to be sure that it guarantees that your append won't have a side-effect on commonTests.

**Pull request merged**